File: rd-net/RdFramework/Impl/HeavySingleContextHandler.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Threading;
using JetBrains.Annotations;
using JetBrains.Collections.Viewable;
using JetBrains.Core;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.Rd.Base;
using JetBrains.Rd.Util;
using JetBrains.Serialization;

#nullable disable

namespace JetBrains.Rd.Impl
{
  internal class HeavySingleContextHandler<T> : RdReactiveBase, ISingleContextHandler<T>
  {
    private readonly ProtocolContexts myHandler;
    private readonly InternRoot<T> myInternRoot;
    private readonly ConcurrentRdSet<T> myProtocolValueSet;

    internal IAppendOnlyViewableConcurrentSet<T> LocalValueSet => myProtocolValueSet;

    public HeavySingleContextHandler(RdContext<T> context, ProtocolContexts handler)
    {
      myHandler = handler;
      Context = context;
      myInternRoot = new InternRoot<T>(context.ReadDelegate, context.WriteDelegate);
      myProtocolValueSet = new ConcurrentRdSet<T>(context.ReadDelegate, context.WriteDelegate, myHandler);
    }

    public object ReadValueBoxed(SerializationCtx context, UnsafeReader reader) => ReadValue(context, reader);

    public RdContextBase ContextBase => Context;

    public RdContext<T> Context { get; }

    protected override void PreInit(Lifetime lifetime, IProtocol proto)
    {
      base.PreInit(lifetime, proto);

      myInternRoot.RdId = proto.Identities.Mix(RdId, "InternRoot");;
      myProtocolValueSet.RdId = proto.Identities.Mix(RdId, "ValueSet");

      myInternRoot.PreBind(lifetime, this, "InternRoot");
      myProtocolValueSet.PreBind(lifetime, this, "ValueSet");
    }

    protected override void Init(Lifetime lifetime, IProtocol proto, SerializationCtx ctx)
    {
      base.Init(lifetime, proto, ctx);
      Assertion.Assert(myHandler.IsSendWithoutContexts,"Must bind context handler without sending contexts to prevent reentrancy");

      myInternRoot.Bind();
      myProtocolValueSet.Bind();

      myProtocolValueSet.View(lifetime, HandleProtocolSetEvent);
    }

    private void HandleProtocolSetEvent(Lifetime lifetime, T value)
    {
      using (myHandler.CreateSendWithoutContextsCookie())
      {
        lifetime.TryBracket(() =>
        {
          myInternRoot.Intern(value);
        }, () =>
        {
          myInternRoot.Remove(value);
        });
      }
    }


    public void WriteValue(SerializationCtx context, UnsafeWriter writer)
    {
      Assertion.Assert(!myHandler.IsSendWithoutContexts, "!myHandler.IsWritingOwnMessages");
      var value = Context.Value;
      if (value == null)
      {
        InternId.Write(writer, InternId.Invalid);
        writer.WriteBoolean(false);
      }
      else
      {
        using (myHandler.CreateSendWithoutContextsCookie())
        {
          AddValueToProtocolValueSetImpl(value);

          var internedId = myInternRoot.Intern(value);
          InternId.Write(writer, internedId);
          if (!internedId.IsValid)
          {
            writer.WriteBoolean(true);
            Context.WriteDelegate(context, writer, value);
          }
        }
      }
    }

    private void AddValueToProtocolValueSetImpl(T value)
    {
      myProtocolValueSet.Add(value);
    }

    public void RegisterValueInValueSet()
    {
      var value = Context.Value;
      if (value == null) return;

      using (myHandler.CreateSendWithoutContextsCookie())
      {
        AddValueToProtocolValueSetImpl(value);
      }
    }

    public T ReadValue(SerializationCtx context, UnsafeReader reader)
    {
      var id = InternId.Read(reader);
      if (!id.IsValid)
      {
        var hasValue = reader.ReadBool();
        if (hasValue)
          return Context.ReadDelegate(context, reader);
        return default;
      }

      return myInternRoot.UnIntern<T>(id);
    }

    public override void OnWireReceived(IProtocol proto, SerializationCtx ctx, UnsafeReader reader, IRdWireableDispatchHelper dispatchHelper)
    {
      var message = "HeavySingleContextHandler can't receive messages";
      Assertion.Fail(message);
    }
  }

  internal class ConcurrentRdSet<T> : RdReactiveBase, IAppendOnlyViewableConcurrentSet<T>, IRdWireable
  {
    private readonly ProtocolContexts myProtocolContexts;
    private readonly ViewableConcurrentSet<T> mySet;
    private readonly ThreadLocal<bool> myIsThreadLocal = new();

    public int Count => mySet.Count;

    public CtxReadDelegate<T> ReadValueDelegate { get; }
    public CtxWriteDelegate<T> WriteValueDelegate { get; }

    public ConcurrentRdSet(CtxReadDelegate<T> readValue, CtxWriteDelegate<T> writeValue, ProtocolContexts protocolContexts, IEqualityComparer<T> comparer = null)
    {
      myProtocolContexts = protocolContexts;
      ReadValueDelegate = readValue;
      WriteValueDelegate = writeValue;
      mySet = new ViewableConcurrentSet<T>(comparer);
      Async = true;
    }

    protected override void PreInit(Lifetime lifetime, IProtocol proto)
    {
      base.PreInit(lifetime, proto);
      proto.Wire.Advise(lifetime, this);
    }

    protected override void Init(Lifetime bindLifetime, IProtocol proto, SerializationCtx ctx)
    {
      View(bindLifetime, (_, value) =>
      {
        if (!myIsThreadLocal.Value) return;

        SendAdd(proto.Wire, SendContext.Of(ctx, value, this));
      });
    }

    private void SendAdd(IWire wire, SendContext<T, ConcurrentRdSet<T>> context)
    {
      wire.Send(RdId, context, static (sendContext, stream) =>
      {
        var value = sendContext.Event;
        const AddRemove kind = AddRemove.Add;
        stream.Write((int)kind);
        sendContext.This.WriteValueDelegate(sendContext.SzrCtx, stream, value);

        ourLogSend.Trace($"{sendContext.This} :: {kind} :: {value.PrintToString()}");
      });
    }

    public override void Print(PrettyPrinter printer)
    {
      base.Print(printer);
      if (!printer.PrintContent) return;

      printer.Print(" [");
      if (Count > 0) printer.Println();

      using (printer.IndentCookie())
      {
        foreach (var v in mySet)
        {
          v.PrintEx(printer);
          printer.Println();
        }
      }
      printer.Println("]");
    }

    public bool Add(T value)
    {
      Assertion.Assert(!myIsThreadLocal.Value);

      myIsThreadLocal.Value = true;
      try
      {
        using (UsingDebugInfo())
        using (myProtocolContexts.CreateSendWithoutContextsCookie())
          return mySet.Add(value);
      }
      finally
      {
        myIsThreadLocal.Value = false;
      }
    }

    public bool Contains(T value) => mySet.Contains(value);
    public void View(Lifetime lifetime, Action<Lifetime, T> action) => mySet.View(lifetime, action);

    public override void OnWireReceived(IProtocol proto, SerializationCtx ctx, UnsafeReader reader, IRdWireableDispatchHelper dispatchHelper)
    {
      var kind = (AddRemove)reader.ReadInt();
      var value = ReadValueDelegate(ctx, reader);
      ReceiveTrace?.Log($"{this} :: {kind} :: {value.PrintToString()}");
      switch (kind)
      {
        case AddRemove.Add:
          mySet.Add(value);
          break;

        case AddRemove.Remove:
          mySet.Remove(value);
          break;

        default:
          throw new ArgumentOutOfRangeException();
      }
    }

    public IEnumerator<T> GetEnumerator() => mySet.GetEnumerator();
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
  }
}

File: rd-kt/rd-framework/src/main/kotlin/com/jetbrains/rd/framework/impl/HeavySingleContextHandler.kt
package com.jetbrains.rd.framework.impl

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.framework.base.IRdBindable
import com.jetbrains.rd.framework.base.IRdWireableDispatchHelper
import com.jetbrains.rd.framework.base.ISingleContextHandler
import com.jetbrains.rd.framework.base.RdReactiveBase
import com.jetbrains.rd.util.*
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.reactive.*
import com.jetbrains.rd.util.string.PrettyPrinter
import com.jetbrains.rd.util.string.print
import com.jetbrains.rd.util.string.printToString

internal class HeavySingleContextHandler<T : Any>(override val context: RdContext<T>, private val contexts: ProtocolContexts) : RdReactiveBase(), ISingleContextHandler<T> {
    private val protocolValueSet = ConcurrentRdSet(context.serializer, contexts)
    @Suppress("UNCHECKED_CAST")
    private val internRoot = InternRoot(context.serializer)

    override fun deepClone(): IRdBindable {
        error("This may not be cloned")
    }

    private fun sendWithoutContexts(block: () -> Unit) {
        contexts.sendWithoutContexts(block)
    }

    val valueSet : IAppendOnlyViewableConcurrentSet<T>
        get() = protocolValueSet

    override fun preInit(lifetime: Lifetime, proto: IProtocol) {
        super.preInit(lifetime, proto)

        protocolValueSet.rdid = proto.identity.mix(rdid, "ValueSet")
        protocolValueSet.preBind(lifetime, this, "ValueSet")

        internRoot.rdid = proto.identity.mix(rdid, "InternRoot")
        internRoot.preBind(lifetime, this, "InternRoot")
    }

    override fun init(lifetime: Lifetime, proto: IProtocol, ctx: SerializationCtx) {
        super.init(lifetime, proto, ctx)

        assert(contexts.isSendWithoutContexts) { "Must bind context handler without sending contexts to prevent reentrancy" }

        protocolValueSet.bind()
        internRoot.bind()

        protocolValueSet.view(lifetime) { valueLifetime, value ->
            handleValueAddedToProtocolSet(valueLifetime, value)
        }
    }

    private fun handleValueAddedToProtocolSet(lifetime: Lifetime, value: T) {
        sendWithoutContexts {
            lifetime.bracketIfAlive({
                internRoot.intern(value)
            }, {
                internRoot.remove(value)
            })
        }
    }

    override fun onWireReceived(proto: IProtocol, buffer: AbstractBuffer, ctx: SerializationCtx, dispatchHelper: IRdWireableDispatchHelper) {
        error("SingleKeyContextHandler does not receive own messages")
    }

    @Suppress("UNCHECKED_CAST")
    override fun writeValue(ctx: SerializationCtx, buffer: AbstractBuffer) {
        assert(!contexts.isSendWithoutContexts) { "Trying to write context with a context-related message, key ${context.key}"}
        val value = context.value
        if(value == null) {
            buffer.writeInternId(InternId.invalid)
            buffer.writeBoolean(false)
        } else {
            sendWithoutContexts {
                addCurrentValueToValueSet(value)

                val internedId = internRoot.intern(value)
                buffer.writeInternId(internedId)
                if (!internedId.isValid) {
                    buffer.writeBoolean(true)
                    context.serializer.write(ctx, buffer, value)
                }
            }
        }
    }

    override fun registerValueInValueSet() {
        val value = context.value ?: return

        sendWithoutContexts {
            addCurrentValueToValueSet(value)
        }
    }

    private fun addCurrentValueToValueSet(value: T) {
        protocolValueSet.add(value)
    }

    override fun readValue(ctx: SerializationCtx, buffer: AbstractBuffer): T? {
        val id = buffer.readInternId()
        return if (!id.isValid) {
            val hasValue = buffer.readBoolean()
            if(hasValue)
                context.serializer.read(ctx, buffer)
            else
                null
        } else
            internRoot.tryUnIntern(id)
    }

    override var async: Boolean
        get() = true
        set(value) = error("SingleKeyContextHandler is always async")

    internal class ConcurrentRdSet<T>(
        private val valueSerializer: ISerializer<T>,
        private val protocolContexts: ProtocolContexts,
    ) : RdReactiveBase(), IAppendOnlyViewableConcurrentSet<T> {


        override var rdid: RdId = RdId.Null

        private val set = ConcurrentViewableSet<T>()
        private val isThreadLocalChange = threadLocalWithInitial { false }

        override val size: Int
            get() = set.size

        init {
            async = true
        }

        override fun preInit(lifetime: Lifetime, proto: IProtocol) {
            super.preInit(lifetime, proto)
            proto.wire.advise(lifetime, this)
        }

        override fun init(lifetime: Lifetime, proto: IProtocol, ctx: SerializationCtx) {
            super.init(lifetime, proto, ctx)
            view(lifetime) { _, value ->
                if (!isThreadLocalChange.get()) return@view

                proto.wire.send(rdid) { writer ->

                    val kind = AddRemove.Add
                    writer.writeEnum(kind)
                    valueSerializer.write(ctx, writer, value)

                    logSend.trace { "set `$location` ($rdid) :: $kind :: ${value.printToString()} "}
                }
            }
        }

        override fun add(value: T): Boolean {
            assert(!isThreadLocalChange.get())

            isThreadLocalChange.set(true)
            try {
                return protocolContexts.sendWithoutContexts {
                    set.add(value)
                }
            } finally {
                isThreadLocalChange.set(false)
            }
        }

        override fun contains(value: T): Boolean {
            return set.contains(value)
        }

        override fun view(lifetime: Lifetime, action: (Lifetime, T) -> Unit) {
            set.view(lifetime, action)
        }

        override fun onWireReceived(proto: IProtocol, buffer: AbstractBuffer, ctx: SerializationCtx, dispatchHelper: IRdWireableDispatchHelper) {
            val kind = buffer.readEnum<AddRemove>()
            val v = valueSerializer.read(ctx, buffer)

            when (kind) {
                AddRemove.Add -> set.add(v)
                AddRemove.Remove -> set.remove(v)
            }
        }

        override fun iterator() = set.iterator()

        override fun print(printer: PrettyPrinter) {
            super.print(printer)
            printer.print(" [")
            if (size > 0) printer.println()

            printer.indent {
                forEach {
                    it.print(printer)
                    printer.println()
                }
            }
            printer.print("]")
        }
    }
}